transport/tcp/serve_tcp_ipmap.go

```diff
@@ -0,0 +1,113 @@
+package tcp
+
+import (
+	"net"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/zrepl/zrepl/transport"
+)
+
+type ipMapEntry struct {
+	ip     net.IP
+	ident  string
+	subnet *net.IPNet
+	zone   string
+}
+
+type ipMap struct {
+	entries []ipMapEntry
+}
+
+func ipMapFromConfig(clients map[string]string) (*ipMap, error) {
+	entries := make([]ipMapEntry, 0, len(clients))
+	for clientInput, clientIdent := range clients {
+		clientIP, clientCIDR, zone := handleIPV6(clientInput)
+
+		if clientCIDR != nil {
+			if len(clientIdent) == 1 && clientIdent == "*" {
+				clientIdent = "willbeclientip"
+			}
+			if len(clientIdent) > 1 && clientIdent[len(clientIdent)-1:] == "*" {
+				clientIdent = strings.TrimRight(clientIdent, "*")
+			}
+		}
+
+		if err := transport.ValidateClientIdentity(clientIdent); err != nil {
+			return nil, errors.Wrapf(err, "invalid client identity for IP %q", clientInput)
+		}
+
+		entries = append(entries, ipMapEntry{
+			ident:  clientIdent,
+			ip:     clientIP,
+			subnet: clientCIDR,
+			zone:   zone,
+		})
+	}
+
+	sort.Sort(bySubnet(entries))
+
+	return &ipMap{entries: entries}, nil
+}
+
+func (m *ipMap) Get(ipAddr *net.IPAddr) (string, error) {
+	for _, e := range m.entries {
+		if e.zone != ipAddr.Zone {
+			continue
+		}
+		if e.ip.Equal(ipAddr.IP) {
+			return e.ident, nil
+		}
+		if e.subnet != nil {
+			contains := e.subnet.Contains(ipAddr.IP)
+			if contains {
+				if e.ident == "willbeclientip" {
+					return ipAddr.IP.String(), nil
+				}
+				return e.ident + ipAddr.IP.String(), nil
+			}
+		}
+	}
+	return "", errors.Errorf("no identity mapping for client IP: %s%%%s", ipAddr.IP, ipAddr.Zone)
+}
+
+func handleIPV6(input string) (net.IP, *net.IPNet, string) {
```

Comment:
a more expressive name would be `parseUserIPMatchExpr` or similar.
Also, I would rather see this return a pointer to a struct, and have the pointer be nil if parsing fails